File: 2022/Day4.lua
function isSubset(x1, x2, y1, y2)
  return x1 <= y1 and y2 <= x2
end

function overlaps(x1, x2, y1, y2)
  return x2 >= y1 and y2 >= x1
end

local count = 0
local oCount = 0

for line in io.lines() do
  local x1, x2, y1, y2 = line:match("(%d+)%-(%d+),(%d+)%-(%d+)")
  x1, x2, y1, y2 = tonumber(x1), tonumber(x2), tonumber(y1), tonumber(y2)

  if isSubset(x1, x2, y1, y2) or isSubset(y1, y2, x1, x2) then
    count = count + 1
  end

  if overlaps(x1, x2, y1, y2) then
    oCount = oCount + 1
  end
end

print(count)
print(oCount)

File: 2022/Day4.cs
using System.Linq;
using System.Collections.Generic;

namespace aoc.Year2022;

public class Day4 : IDay {
  public void PartOne(string input) {
    var pairs = GeneratePairs(input);

    var hasSubset = pairs.Where(p =>
      Helpers.IsSubset(p[0].From, p[0].To, p[1].From, p[1].To)
      || Helpers.IsSubset(p[1].From, p[1].To, p[0].From, p[0].To));

    Console.WriteLine(hasSubset.Count());
  }

  public void PartTwo(string input) {
    var pairs = GeneratePairs(input);

    var hasOverlaps = pairs.Where(p => Helpers.Overlaps(p[0].From, p[0].To, p[1].From, p[1].To));

    Console.WriteLine(hasOverlaps.Count());
  }

  private static IEnumerable<List<(int From, int To)>> GeneratePairs(string input) {
    return input
      .ToLines()
      .Select(l => l
        .Split(",")
        .Select(x => x
          .Split("-")
          .Select(int.Parse)
          .ToList())
        .Select(x => (From: x[0], To: x[1]))
        .ToList());
  }
}
